### ISSUE TITLE
Add new lintr to ensure subsetting is done with integer

### ISSUE THREAD

**Issue by TanguyBarthelemy** [r-lib/lintr#2625]
Is it possible to add a new lintr to ensure that the subsetting of a vector, data.frame or matrix is done using integers?

For example, the following code will be marked by {lintr}:

```r
letters[c(1, 2, 10)]
iris[1:3, 3]
```

and the suggested correction would be:

```r
letters[c(1L, 2L, 10L)]
iris[1:3, 3L]
```

**Comment by MichaelChirico** [comment]
That should be covered by `implicit_integer_linter()`.

**Comment by TanguyBarthelemy** [comment]
When I call `lintr::lint_dir()` or `lintr::lint_package()`, I don't get a marker for `implicit_integer_linter`.
So I changed my `.lintr` file in:

```
linters: linters_with_defaults(
        indentation_linter = lintr::indentation_linter(indent = 4L),
        implicit_integer_linter = lintr::implicit_integer_linter(allow_colon = FALSE)
    )
encoding: "UTF-8"
```

Are there other lintr not automatically applied and how can they be applied automatically?

Ok I didn't know that the recommendation was to change `1:3` to `1L:3L` because both are the same object.

Do you have a guide or manual that lists these best practices?

Thank you very much!

**Comment by MichaelChirico** [comment]
> Ok I didn't know that the recommendation was to change 1:3 to 1L:3L because both are the same object.

I think it's up to taste, personally I prefer `1:3`.

I confirm the `implicit_integer_linter()` WAI:

```r
lintr::lint(linters = lintr::implicit_integer_linter(), text = "
letters[c(1, 2, 10)]
iris[1:3, 3]
")
```

Output:

```
<text>:2:12: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
letters[c(1, 2, 10)]
          ~^
<text>:2:15: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
letters[c(1, 2, 10)]
             ~^
<text>:2:19: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
letters[c(1, 2, 10)]
                ~~^
<text>:3:7: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
iris[1:3, 3]
     ~^
<text>:3:9: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
iris[1:3, 3]
       ~^
<text>:3:12: style: [implicit_integer_linter] Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles. Disable this with a comment: `# nolint` (one line) or `# lintr: disable` (multiple lines). See go/tidy-style#linting for more.
iris[1:3, 3]
          ~^
```

It's not a default linter:

https://github.com/r-lib/lintr/blob/48b36ce3a281c12368af0c5a381a4681c2a00f9d/inst/lintr/linters.csv?plain=1#L44

> Do you have a guide or manual that lists these best practices?

Closest we get is the tidyverse guide:

https://style.tidyverse.org/

You can also see the `best_practices` tag among linters:

```r
names(lintr::linters_with_tags("best_practices"))
#  [1] "absolute_path_linter"           "any_duplicated_linter"          "any_is_na_linter"               "boolean_arithmetic_linter"      "class_equals_linter"           
#  [6] "commented_code_linter"          "condition_message_linter"       "conjunct_test_linter"           "cyclocomp_linter"               "empty_assignment_linter"       
# [11] "expect_comparison_linter"       "expect_length_linter"           "expect_named_linter"            "expect_not_linter"              "expect_null_linter"            
# [16] "expect_s3_class_linter"         "expect_s4_class_linter"         "expect_true_false_linter"       "expect_type_linter"             "extraction_operator_linter"    
# [21] "fixed_regex_linter"             "for_loop_index_linter"          "function_argument_linter"       "function_return_linter"         "ifelse_censor_linter"          
# [26] "implicit_assignment_linter"     "implicit_integer_linter"        "is_numeric_linter"              "length_levels_linter"           "lengths_linter"                
# [31] "library_call_linter"            "literal_coercion_linter"        "nonportable_path_linter"        "outer_negation_linter"          "paste_linter"                  
# [36] "redundant_equals_linter"        "redundant_ifelse_linter"        "regex_subset_linter"            "routine_registration_linter"    "scalar_in_linter"              
# [41] "seq_linter"                     "sort_linter"                    "system_file_linter"             "T_and_F_symbol_linter"          "undesirable_function_linter"   
# [46] "undesirable_operator_linter"    "unnecessary_lambda_linter"      "unnecessary_nested_if_linter"   "unnecessary_placeholder_linter" "unreachable_code_linter"       
# [51] "unused_import_linter"           "vector_logic_linter"            "yoda_test_linter"  
```

Or just turn on all linters and selectively disable them (my own preferred approach for my packages): `lintr::all_linters()`.
